File: src/main/java/de/mikeyllp/miniGamesV4/games/hideandseek/utils/WaitingForPlayersUtils.java
package de.mikeyllp.miniGamesV4.games.hideandseek.utils;

import net.kyori.adventure.text.minimessage.MiniMessage;
import org.bukkit.Sound;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.entity.Player;
import org.bukkit.plugin.java.JavaPlugin;
import org.bukkit.scheduler.BukkitRunnable;

import java.util.ArrayList;
import java.util.List;

import static de.mikeyllp.miniGamesV4.games.hideandseek.HideAndSeekGame.startGame;
import static de.mikeyllp.miniGamesV4.games.hideandseek.storage.HideAndSeekGameGroups.listUntilX;
import static de.mikeyllp.miniGamesV4.games.hideandseek.utils.formatTimeUtils.formatTimer;

public class WaitingForPlayersUtils {
    public static BukkitRunnable waitingTask;

    public static final List<Integer> timerList = new ArrayList<>();
    private static boolean timerRunning = false;

    public static void startWaitingTask(JavaPlugin plugin) {
        FileConfiguration config = plugin.getConfig();
        int rawTimer = config.getInt("timeAutoStartHASGroup");
        int maxPlayers = config.getInt("maxPlayersPerHASGroup");
        int minPlayers = config.getInt("minPlayersPerHASGroup");
        MiniMessage mm = MiniMessage.miniMessage();

        // Create a new task
        waitingTask = new BukkitRunnable() {
            // This is for the timer countdown
            boolean countdownStarted = false;
            int timer = rawTimer;

            @Override
            public void run() {
                int currentSize = listUntilX.size();
                // Check if the there are no players in the list if so cancel the task
                if (currentSize == 0) {
                    cancel();
                    waitingTask = null;
                    timerRunning = false;
                    countdownStarted = false;
                    return;
                }
                // Checks if enough players are in the list, if not cancel the timer
                if (listUntilX.size() < minPlayers && timerRunning) {
                    for(Player p : listUntilX) {
                        p.playSound(p.getLocation(), Sound.UI_BUTTON_CLICK, 1f, 1f);
                    }
                    timerRunning = false;
                    countdownStarted = false;
                    int timer = rawTimer;
                    return;
                }
                // If the countdown has not started and enough players are in the list, start the countdown
                if (!countdownStarted && currentSize >= minPlayers) {
                    countdownStarted = true;
                    timerRunning = true;
                    int timer = rawTimer;
                }

                // This is the countdown logic, if the countdown has started, send the action bar message
                if (countdownStarted) {
                    String showTimer = formatTimer(timer);
                    for (Player p : listUntilX) {
                        p.sendActionBar(mm.deserialize("<gold>Spiel startet in: <color:#00E5E5>" + showTimer +
                                "</color> (<color:#00E5E5>" + currentSize + "</color>/<color:#00E5E5>" + maxPlayers + "</color>)</gold>"));
                        if (timer <= 5) {
                            p.playSound(p.getLocation(), Sound.BLOCK_NOTE_BLOCK_HAT, 1f, 1f);
                        }
                    }
                    if (timer <= 0 || currentSize >= maxPlayers) {
                        for (Player p : listUntilX) {
                            p.playSound(p.getLocation(), Sound.ENTITY_PLAYER_LEVELUP, 1.0F, 1.5F);
                        }
                        startGame(plugin, config);
                        cancel();
                        waitingTask = null;
                        timerRunning = false;
                    }
                    timer--;
                } else {
                    for (Player p : listUntilX) {
                        p.sendActionBar(mm.deserialize("<gold>Warten auf weitere Spieler... (<color:#00E5E5>" + currentSize + "</color>/<color:#00E5E5>" + maxPlayers + "</color>)</gold>"));
                    }
                }
            }
        };
        waitingTask.runTaskTimer(plugin, 0L, 20L);
    }
}

File: src/main/java/de/mikeyllp/miniGamesV4/games/hideandseek/utils/WaitingForPlayersUtils.kt
package de.mikeyllp.miniGamesV4.games.hideandseek.utils

import de.mikeyllp.miniGamesV4.games.hideandseek.HideAndSeekGame
import de.mikeyllp.miniGamesV4.games.hideandseek.storage.HideAndSeekGameGroups
import de.mikeyllp.miniGamesV4.messages.Translator
import de.mikeyllp.miniGamesV4.plugin
import org.bukkit.Sound
import org.bukkit.scheduler.BukkitRunnable

object WaitingForPlayersUtils {
    var waitingTask: BukkitRunnable? = null

    val timerList: MutableList<Int> = ArrayList<Int>()
    private var timerRunning = false

    fun startWaitingTask() {
        val config = plugin.getConfig()
        val rawTimer = config.getInt("timeAutoStartHASGroup")
        val maxPlayers = config.getInt("maxPlayersPerHASGroup")
        val minPlayers = config.getInt("minPlayersPerHASGroup")



        waitingTask = object : BukkitRunnable() {

            var countdownStarted: Boolean = false
            var timer: Int = rawTimer

            override fun run() {
                val currentSize: Int = HideAndSeekGameGroups.listUntilX.size

                if (currentSize == 0) {
                    cancel()
                    waitingTask = null
                    timerRunning = false
                    countdownStarted = false
                    return
                }

                if (HideAndSeekGameGroups.listUntilX.size < minPlayers && timerRunning) {
                    for (p in HideAndSeekGameGroups.listUntilX) {
                        p.playSound(p.location, Sound.UI_BUTTON_CLICK, 1f, 1f)
                    }
                    timerRunning = false
                    countdownStarted = false
                    val timer = rawTimer
                    return
                }

                if (!countdownStarted && currentSize >= minPlayers) {
                    countdownStarted = true
                    timerRunning = true
                    val timer = rawTimer
                }

                if (countdownStarted) {
                    val showTimer = FormatTimeUtils.formatTimer(timer)
                    for (p in HideAndSeekGameGroups.listUntilX) {

                        p.sendActionBar(
                            Translator.translatable(
                                "has.action-bar.waiting-for-players.timer",
                                showTimer,
                                currentSize.toString(),
                                maxPlayers.toString()
                            )
                        )

                        if (timer <= 5) {
                            p.playSound(p.location, Sound.BLOCK_NOTE_BLOCK_HAT, 1f, 1f)
                        }
                    }

                    if (timer <= 0 || currentSize >= maxPlayers) {
                        for (p in HideAndSeekGameGroups.Companion.listUntilX) {
                            p.playSound(p.location, Sound.ENTITY_PLAYER_LEVELUP, 1.0f, 1.5f)
                        }
                        HideAndSeekGame.startGame()
                        cancel()
                        waitingTask = null
                        timerRunning = false
                    }
                    timer--
                } else {
                    for (p in HideAndSeekGameGroups.listUntilX) {
                        p.sendActionBar(
                            Translator.translatable(
                                "has.action-bar.waiting-for-players.no-timer",
                                currentSize.toString(),
                                maxPlayers.toString(),
                            )
                        )
                    }
                }
            }
        }
        waitingTask!!.runTaskTimer(plugin, 0L, 20L)
    }
}
